### PR TITLE
Make token optional in all HfApi methods.

### PR DESCRIPTION
**Pull request by sgugger**
This PR makes the `token` argument of all HfApi methods optional and default to the saved token. As you can see in the updated docs, the result is quite nice in the sense a user never has to retrieve their token (and thus get some chance of leaking it by mistake).

Since this changes the order of the argument and is thus a breaking change, some internal magic is implemented to detect if the argument at the old place of the token is the token, when the argument at the place of the token is not a valid token, and then do some switch.

### CODE REVIEW

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -328,6 +328,10 @@ def whoami(self, token: Optional[str] = None) -> Dict:
         """
         if token is None:
             token = HfFolder.get_token()
+        if token is None:
+            raise ValueError(
+                "You need to pass a valid `token` or login by using `huggingface-cli login`"
+            )
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:334]: Yes, very nice improvement

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -563,10 +588,17 @@ def dataset_info(
         d = r.json()
         return DatasetInfo(**d)
 
+    def _is_valid_token(self, token: str):
+        try:
+            self.whoami(token=token)
+            return True
+        except HTTPError:
+            return False
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:596]: Nice helper, would it be possible to add a bit of documentation?

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -591,6 +623,23 @@ def create_repo(
...
+                    "`huggingface-cli login`."
+                )
+        elif not self._is_valid_token(token):
+            if self._is_valid_token(name):
+                warnings.warn(
+                    "`create_repo` now takes `token` as an optional positional argument. "
+                    "Be sure to adapt your code!",
+                    FutureWarning,
+                )
+                token, name = name, token
```
Comment (LysandreJik) [inline src/huggingface_hub/hf_api.py:640]: Perfect!

**Review by LysandreJik — COMMENTED** [review]
This is a very welcome change which should make many things clearer. I think having the `token` as first parameter in most methods isn't ideal, as it is generally not the key parameter of the method.

Ont nit, however: would it be possible to split the PR in two, with one PR changing the API and the second changing the tests? Modifying both at the same time is a bit dangerous as we try to prevent breaking changes.

@osanseviero, this PR introduces a change in the methods', but Sylvain also introduced mechanisms to be backwards compatible as it would otherwise be very breaking. This will, however, change in a future version with future warnings removed.

Do you know if libraries have an unpinned version of `huggingface_hub`, and, if so, if they use any of the changed methods here? Namely, `create_repo`, `delete_repo`, `update_repo_visibility`, and `upload_file`? Thank you.

**Comment by LysandreJik** [comment]
Also cc @stevhliu, could you link which documentation pages should be updated to reflect this change? Thanks a lot

**Comment by stevhliu** [comment]
I think these are the only two pages that will need to be updated:

- [Integrate a library with the Hub
](https://huggingface.co/docs/hub/adding-a-library)
- [How to integrate upstream utilities in your library](https://huggingface.co/docs/hub/how-to-upstream)

Let me know if I can help update these :)

**Comment by sgugger** [comment]
I've already updated those to the new API, normally.

**Comment by osanseviero** [comment]
Sorry for the delay with my response. Thanks for this PR @sgugger!!

I think there are 4 actionable repos (1 officially supported, 3 non official)

**Officially supported libraries**

* I think we definitively need to change this (unpinned, not using keyword args)
https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/SentenceTransformer.py#L437-L458

* AllenNLP is fine but we can simplify the code with this PR in a later stage
https://github.com/allenai/allennlp/blob/main/allennlp/common/push_to_hf.py#L136-L150

* spaCy is also fine, and again we can simplify it with this change
https://github.com/explosion/spacy-huggingface-hub/blob/main/spacy_huggingface_hub/push.py#L81-L89


**Non official**

* Glasses https://github.com/FrancescoSaverioZuppichini/glasses/blob/master/glasses/utils/weights/storage/hubs/HFModelHub.py#L202-L210
* HuggingTweets https://github.com/borisdayma/huggingtweets/blob/8a4cdefd8b7ef9fef56c76190b5afb97b444108b/huggingtweets-demo.ipynb
* HuggingArtitsts https://github.com/AlekseyKorshuk/huggingartists/blob/master/huggingartists-demo.ipynb

**Comment by LysandreJik** [comment]
Thank you for doing the research @osanseviero!

@Narsil this touches to some of the code in api-inference-community, could you take a quick look at the changes?

**Review by LysandreJik — APPROVED** [review]
LGTM! As seen with @sgugger and @osanseviero, the only library that uses an unpinned version of `huggingface_hub` and that would potentially break is `sentence-transformers` cc @nreimers.

The workarounds in `create_repo` should prevent any breaking change, however.

**Pull request merged**